src/Compilers/Core/MSBuildTask/DiscoverEditorConfigFiles.cs

```diff
@@ -0,0 +1,164 @@
+                    {
+                        directoriesToVisit.Enqueue(directory);
+                    }
+                }
+            }
+
+            foreach (var inputFile in InputFiles)
+            {
+                addNewDirectoryIfNotAlreadyAdded(new DirectoryInfo(Path.GetDirectoryName(Path.GetFullPath(inputFile.ItemSpec))));
+            }
+
+            while (directoriesToVisit.Count > 0)
+            {
+                var directory = directoriesToVisit.Dequeue();
+                var editorConfigFilePath = Path.Combine(directory.FullName, ".editorconfig");
+                bool isRootEditorConfig = false;
+
+                try
+                {
+                    if (File.Exists(editorConfigFilePath))
+                    {
+                        editorConfigFiles.Add(new TaskItem(editorConfigFilePath));
+
+                        // We will attempt to determine if this is a root .editorconfig, so we can stop enumerating farther if necessary.
+                        // We will do no further interpretation of the .editorconfig.
+
+                        isRootEditorConfig = FileIsRootEditorConfig(editorConfigFilePath);
+                    }
+                }
+                catch (IOException exception)
+                {
+                    // We'll log the error and continue; the "it failed" is handled in Execute().
+                    Log.LogErrorFromResources(nameof(ErrorString.General_UnableToReadFile), editorConfigFilePath, exception.Message);
+                }
+
+                if (!isRootEditorConfig && directory.Parent != null)
+                {
+                    addNewDirectoryIfNotAlreadyAdded(directory.Parent);
+                }
+            }
+
+            EditorConfigFiles = editorConfigFiles.ToArray();
+        }
+
+        // These Regexes are the regexes from Microsoft.VisualStudio.CodingConventions.EditorConfig.
+        private static readonly Regex s_sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        private static readonly Regex s_propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        internal static bool FileIsRootEditorConfig(string editorConfigFilePath)
+        {
+            using (FileStream fileStream = OpenFile(editorConfigFilePath))
+            using (StreamReader reader = new StreamReader(fileStream))
+            {
+                string line;
+                while ((line = reader.ReadLine()) != null)
+                {
+                    // Once we're in a section, we don't need to go looking for root = true anymore
+                    if (s_sectionMatcher.IsMatch(line))
+                    {
+                        return false;
+                    }
+
+                    var propMatches = s_propertyMatcher.Matches(line);
+                    if (propMatches.Count == 1 && propMatches[0].Groups.Count > 2)
```

Comment:
Please document what these groups actually mean. Maybe my regex-fu is failing me, but there should only be 3 capture groups in this regex, so why are you checking for greater than 2 instead of just == 3? I see that groups 2 and 3 are given names before, but as far as I can tell group 1 is never used so I have no idea what it is.